core/trino-main/src/main/java/io/trino/connector/InternalMetadataProvider.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.connector;
+
+import io.trino.FullConnectorSession;
+import io.trino.Session;
+import io.trino.metadata.MaterializedViewDefinition;
+import io.trino.metadata.MetadataManager;
+import io.trino.metadata.QualifiedObjectName;
+import io.trino.metadata.TableHandle;
+import io.trino.metadata.ViewDefinition;
+import io.trino.spi.connector.CatalogSchemaTableName;
+import io.trino.spi.connector.ColumnMetadata;
+import io.trino.spi.connector.ConnectorRelationMetadata;
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.connector.MetadataProvider;
+
+import java.util.Optional;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public class InternalMetadataProvider
+        implements MetadataProvider
+{
+    private final MetadataManager metadataManager;
+
+    public InternalMetadataProvider(MetadataManager metadataManager)
+    {
+        this.metadataManager = requireNonNull(metadataManager, "metadataManager is null");
+    }
+
+    @Override
+    public Optional<ConnectorRelationMetadata> getRelationMetadata(ConnectorSession connectorSession, CatalogSchemaTableName tableName)
+    {
+        Session session = ((FullConnectorSession) connectorSession).getSession();
```

Comment:
Is that a safe cast?

Reply:
Yes, i think so. You can check the codebase for other examples.